lrauv_system_tests/include/lrauv_system_tests/Subscription.hh

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/*
+ * Development of this module has been funded by the Monterey Bay Aquarium
+ * Research Institute (MBARI) and the David and Lucile Packard Foundation
+ */
+
+#ifndef LRAUV_SYSTEM_TESTS__SUBSCRIPTION_HH
+#define LRAUV_SYSTEM_TESTS__SUBSCRIPTION_HH
+
+#include <deque>
+#include <mutex>
+#include <string>
+
+#include <ignition/transport/Node.hh>
+
+namespace lrauv_system_tests
+{
+
+/// A stateful wrapper for Ignition Transport subscriptions.
+template<typename MessageT>
+class Subscription
+{
+  /// Default constructor.
+  public: Subscription() = default;
+
+  /// Subscribe to a topic.
+  /// \param[in] _node Node to use to subscribe.
+  /// \param[in] _topicName Name of the topic to subscribe.
+  /// \param[in] _messageHistoryDepth Maximum size for
+  /// message history buffer. Buffer grows indefinitely
+  /// by default.
+  public: void Subscribe(
+      ignition::transport::Node &_node,
+      const std::string &_topicName,
+      size_t _messageHistoryDepth = 0u)
+  {
+    if (this->subscribed)
+    {
+      throw std::logic_error("cannot ");
```

Comment:
I think the error message got cut short. Also might be worth making this the constructor rather than having a separate `Subscribe()`.

Reply:
> I think the error message got cut short.

Yeap! Good catch. Fixed in https://github.com/osrf/lrauv/pull/195/commits/34cfeeba428b0b0e175976fc0dba48f8df6753a6.

>  Also might be worth making this the constructor rather than having a separate `Subscribe()`.

About that. I agree with you, I don't like the half constructed object. However, depending on how you write tests, having a default constructor can be useful. E.g. [here](https://github.com/osrf/lrauv/pull/192/files#diff-679ef17c89eb7e8b08f01621e4846008baafc06f4100bd53e70d4af5cbf0a53bR288-R289), I'd have to cram everything in a field initializer or dynamically allocate the object to defer construction if there wasn't a `Subscribe()`.

Reply:
Fair enough. Makes sense.